types/decimal.go

```diff
@@ -0,0 +1,351 @@
+		d = d.Neg(d)
+		chopped.Neg(chopped)
+		return chopped
+	}
+
+	// get the trucated quotient and remainder
+	quo, rem := big.NewInt(0), big.NewInt(0)
+	quo, rem = quo.QuoRem(d, precisionReuse, rem)
+
+	if rem.Sign() == 0 { // remainder is zero
+		return quo
+	}
+
+	lenWhole := len(d.String())
+	if quo.Sign() == 0 { // only the decimal places (ex. 0.1234)
+		lenWhole++
+	}
+	lenQuo := len(quo.String())
+	lenRem := len(rem.String())
+	leadingZeros := lenWhole - (lenQuo + lenRem) // leading zeros removed from the remainder
+
+	zerosToAdd := int64(lenRem - 1 + leadingZeros)
+	multiplier := new(big.Int).Exp(big.NewInt(10), big.NewInt(zerosToAdd), nil)
+	fiveLine := new(big.Int).Mul(big.NewInt(5), multiplier)
+
+	switch rem.Cmp(fiveLine) {
+	case -1:
+		chopped = quo
+		return
+	case 1:
+		chopped = quo.Add(quo, big.NewInt(1))
+		return
+	default: // bankers rounding must take place
+		// always round to an even number
+		if quo.Bit(0) == 0 {
+			chopped = quo
+			return
+		}
+		chopped = quo.Add(quo, big.NewInt(1))
+		return
+	}
+}
+
+// RoundInt64 rounds the decimal using bankers rounding
+func (d Dec) RoundInt64() int64 {
+	if !d.Int.IsInt64() {
+		panic("Int64() out of bound")
+	}
+	return BankerRoundChop(d.Int, Precision).Int64()
+}
+
+// RoundInt round the decimal using bankers rounding
+func (d Dec) RoundInt() Int {
+	return NewIntFromBigInt(BankerRoundChop(d.Int, Precision))
+}
+
+//___________________________________________________________________________________
+
+// wraps d.MarshalText()
+func (d Dec) MarshalAmino() (string, error) {
+	if d.Int == nil {
+		d.Int = new(big.Int)
+	}
+	bz, err := d.Int.MarshalText()
+	return string(bz), err
+}
+
+// requires a valid JSON string - strings quotes and calls UnmarshalText
+func (d *Dec) UnmarshalAmino(text string) (err error) {
+	tempInt := new(big.Int)
+	err = tempInt.UnmarshalText([]byte(text))
+	if err != nil {
+		return err
+	}
+	d.Int = tempInt
+	return nil
+}
+
+// MarshalJSON defines custom encoding scheme
+func (d Dec) MarshalJSON() ([]byte, error) {
+	if d.Int == nil {
+		d.Int = new(big.Int)
+	}
+
+	text, err := d.Int.MarshalText()
+	if err != nil {
+		return nil, err
+	}
+	return json.Marshal(string(text))
+}
+
+// UnmarshalJSON defines custom decoding scheme
+func (d *Dec) UnmarshalJSON(bz []byte) error {
+	if d.Int == nil {
+		d.Int = new(big.Int)
+	}
+
+	var text string
+	err := json.Unmarshal(bz, &text)
+	if err != nil {
+		return err
+	}
+	return d.Int.UnmarshalText([]byte(text))
+}
+
+//___________________________________________________________________________________
+// helpers
+
+// test if two decimal arrays are equal
+func DecsEqual(d1s, d2s []Dec) bool {
+	if len(d1s) != len(d2s) {
+		return false
+	}
+
+	for i, d1 := range d1s {
+		if !d1.Equal(d2s[i]) {
+			return false
+		}
+	}
+	return true
+}
+
+// intended to be used with require/assert:  require.True(DecEq(...))
+func DecEq(t *testing.T, exp, got Dec) (*testing.T, bool, string, Dec, Dec) {
```

Comment:
Lol whoops I forgot there was a reason why this was here - you can't reference functions you've defined in a `xxx_test.go` file outside of that package - it needs to in some kind of a utilities file if we want to separate it out (like `test_common.go`)... since this is only one small function I'm going to leave it where it was for now